CMS OAuth Server blog post (#4141)
pulumi/docs: content/blog/deploying-the-infrastructure-of-oauth-server-for-cms-app/index.md

For the Authorization callback URL we should enter `https://{{the domain of your OAuth App}}/callback/github` which is specified by the main.go.

```go
    github.New(
        os.Getenv("GITHUB_KEY"), os.Getenv("GITHUB_SECRET"),
        // concatenate with the host name
        fmt.Sprintf("%s/callback/github", host),
        "public_repo",
    )
```

If you are using different storaging system like Bitbucket and Gitlab, you can specify callback url to be `https://{{the domain of your OAuth App}}/callback/{{ backend name }}` and relace `{{backend name}}` with `bitbucket` or `github` correspondingly.

Then we can copy and paste the github key and secret as stack configuration as well.

```bash
$ pulumi config set netlify-cms-oauth-provider-infrastructure:githubKey {{YOUR_GITHUB_KEY}}
$ pulumi config set --secret netlify-cms-oauth-provider-infrastructure:githubSecret
$ {{YOUR_GITHUB_SECRET}}
```

`--secret` would safely encrypt your github secret. Also,
don't directly append the secret to the command like this `$ pulumi config set --secret netlify-cms-oauth-provider-infrastructure:githubKey {{YOUR_GITHUB_SECRET}}`.

> Because it might cause the secret to being stored inside the command memory. We should only specify the key without a name and hit ENTER key, then you can type secret on next line(won't show the value)

Reviewer: ```suggestion
Your secret would be stored in the command history. Instead, specify the key without a value and enter the secret at the prompt which hides the value.
```

> For `TARGET_PORT` that is how we passed in the `targetGroupPort` configuration value to main.go. That is also the port number that we want the server itself to serve. We also checked if this environment variable is set in main.go in case we need to test main.go locally without Pulumi.

Reviewer: ```suggestion
We pass `TARGET_PORT` to `targetGroupPort` which is the port used to serve the application. We also check if this variable is set in main.go in case we need to test locally without Pulumi.
```

```go
targetGroupPort := os.Getenv("TARGET_PORT")
    if targetGroupPort == "" {
        targetGroupPort = "80"
    }
    listenPort := ":" + targetGroupPort
    fmt.Print("Started running on", listenPort, "\n")
    // listen on port 80 where we created the target group
    fmt.Println(http.ListenAndServe(listenPort, nil))
```

> As mentioned previously, `GITHUB_SCOPE` is the environment variable to tell main.go what kind of access CMS needs to have on the target repository. It is also an optional stack configuration where the default is set as public_repo in the main.go. We can specify multiple scopes by a separate scope with commas.

Reviewer: ```suggestion
As mentioned previously, `GITHUB_SCOPE` sets the type of access the CMS has for the target repository. It is also an optional stack configuration where the default is set as `public_repo`.
```

### Create a Certificate for the OAuth Server

> The process is similar how we created certificate for CMS in our [previous post]({{< relref "/blog/deploying-netlify-cms-on-aws" >}}).

Reviewer: ```suggestion
The process is similar to how we created a certificate for CMS in our [previous post]({{< relref "/blog/deploying-netlify-cms-on-aws" >}}).
```

> Similarly, we also referenced Pulumi's [static website example](https://github.com/pulumi/examples/tree/master/aws-ts-static-website).

Reviewer: ```suggestion
```

> We can extract what we need just for creating the certificate and delete irrelevant parts including uploading content to the S3 bucket and configure CloudFront. Then the rest only involves creating an east region provider, a certificate, and certificate validation.

Reviewer: ```suggestion
We use just what we need just for creating the certificate.
```

Author: I think I should still point out which part of the example that is irrelevant/ relevant and show the way to use Pulumi examples

Reviewer 2: @zephyrz73 I do see your point about wanting to highlight that we can easily just take the relevant parts from the other infrastructure app. :) However, I agree that the suggestion made by @spara captures the same thing you are saying but in a succinct way and does imply that we are taking what we need just for creating the certificate itself. I recommend applying the suggestion here.

> However, for the part of creating Alias Record, the values of the alias field should be filled with information about the application load balancer that we created.

Reviewer: ```suggestion
When we create the Alias Record, we configure it with application load balancer settings.
```

```typescript
// Creates a new Route53 DNS record pointing the domain to the CloudFront distribution.
function createAliasRecord(
    targetDomain: string, lb: awsx.elasticloadbalancingv2.ApplicationLoadBalancer): aws.route53.Record {
    const domainParts = getDomainAndSubdomain(targetDomain);
    const hostedZoneId = aws.route53.getZone({ name: domainParts.parentDomain }, { async: true }).then(zone => zone.zoneId);
    return new aws.route53.Record(
        targetDomain,
        {
            name: domainParts.subdomain,
            zoneId: hostedZoneId,
            type: "A",
            aliases: [
                {
                    name: lb.loadBalancer.dnsName,
                    zoneId: lb.loadBalancer.zoneId,
                    evaluateTargetHealth: true,
                },
            ],
        });
}
// Create the aliasRecord with targetdomain and application load balancer
const aRecord = createAliasRecord(cmsStackConfig.targetDomain, alb);
```

> Then we could run `pulumi up` to deploy all of this.

Reviewer: ```suggestion
We run `pulumi up` to deploy.
```

### Github Workflow (Optional)

> Similar to the cms folder we have to build the Github workflow under folder `cms-oauth/.github/workflows/build-and-deploy.yml`. The workflow is also using the repository secret that we set in the Github.

Reviewer: ```suggestion
We can automate our build with the Github workflow in the folder `cms-oauth/.github/workflows/build-and-deploy.yml`. The workflow uses the repository secret we set in Github.
```

![Github Secret](./github_secret.jpg)

## Last Step

> Don't forget to update the CMS configuration file.

Reviewer: ```suggestion
Finally, we need to update the CMS configuration file.
```

> To put all of our deployment of CMS and CMS OAuth Server together, we should specify the site_domain and URL in `./public/config.yml` of the cms folder.

Reviewer: ```suggestion
To combine our deployment of CMS and CMS OAuth Server, we specify the site_domain and URL in `./public/config.yml` of the CMS folder.
```

```yaml
backend:
  name: github
  # Replace this with the Github repo you want to make change
  repo: github-username/target-github-repo
  # This site_domain and base_url are sudo domains
  # Replace site_domain with targetDomain pulumi stack configuration inside cms/infrastructure folder
  # Replace base_url with targetDomain pulumi stack configuration inside cms-oauth/infrastructure folder
  site_domain: https://some-cms-domain.pulumi-demos.com
  base_url: https://some-oauth-domain.pulumi-demos.com
```

> The site_domain is the domain of the CMS and the base_url is the domain of OAuth Server that we deployed. They should be the same as the `targetDomain` variable we set in the stack configuration of both CMS and OAuth Server.

Reviewer: ```suggestion
The site_domain is the URL of the CMS and the base_url is the URL of the OAuth Server . They should be the same as the `targetDomain` variable we set in the stack configuration of both CMS and OAuth Server.
```

                },
                {
                    name: "GITHUB_SCOPE",
                    value: inputGithubScope
                },
            ]
        },
    },
    desiredCount: 1,
});
```

For `HOST`, we set the URL for the OAuth Server by running `pulumi config set pulumi-website-cms:targetDomain https://some-cms-oauth-domain.pulumi-demos.com` and replace the URL.

`SESSION_SECRET` is another environment variable passed to main.go that we generated by using Pulumi's `random.RandomPassword`.

To get `GITHUB_SECRET` and `GITHUB_TOKEN` credentials, we need to register the Github OAuth application. Netlify provides [instructions](https://docs.netlify.com/visitor-access/oauth-provider-tokens/#setup-and-settings) for obtaining these credentials.

The Home Page URL is the targetDomain from our [previous post]({{< relref "/blog/deploying-netlify-cms-on-aws" >}}).

The Authorization callback URL is `https://{{the domain of your OAuth App}}/callback/github`.

```go
    github.New(
        os.Getenv("GITHUB_KEY"), os.Getenv("GITHUB_SECRET"),
        // concatenate with the host name
        fmt.Sprintf("%s/callback/github", host),
        "public_repo",
    )
```

> If you use different repositories, like Bitbucket and Gitlab, you can specify callback URL to `https://{{the domain of your OAuth App}}/callback/{{ backend name }}` and relace `{{backend name}}` with `bitbucket` or `github` depending upon the repository.

Reviewer 2: ```suggestion
If you use different repositories, like Bitbucket and Gitlab, you can specify callback URL to `https://{{the domain of your OAuth App}}/callback/{{ backend name }}` and replace `{{backend name}}` with `bitbucket` or `github` depending upon the repository.
```

We can set the github key and secret in the stack configuration.

```bash
$ pulumi config set netlify-cms-oauth-provider-infrastructure:githubKey {{YOUR_GITHUB_KEY}}
$ pulumi config set --secret netlify-cms-oauth-provider-infrastructure:githubSecret
$ {{YOUR_GITHUB_SECRET}}
```

> The `--secret` flag encrypts your Github secret.

Reviewer 2: ```suggestion
The `--secret` flag encrypts your GitHub secret.
```

        targetDomain,
        {
            name: domainParts.subdomain,
            zoneId: hostedZoneId,
            type: "A",
            aliases: [
                {
                    name: lb.loadBalancer.dnsName,
                    zoneId: lb.loadBalancer.zoneId,
                    evaluateTargetHealth: true,
                },
            ],
        });
}
// Create the aliasRecord with targetdomain and application load balancer
const aRecord = createAliasRecord(cmsStackConfig.targetDomain, alb);
```

We run `pulumi up` to deploy.

### Github Workflow (Optional)

We can automate our build with the Github workflow in the folder `cms-oauth/.github/workflows/build-and-deploy.yml`. The workflow uses the repository secret we set in Github.

![Github Secret](./github_secret.jpg)

## Last Step

Finally, we need to update the CMS configuration file.

To combine our deployment of CMS and CMS OAuth Server, we specify the site_domain and URL in `./public/config.yml` of the CMS folder.

```yaml
backend:
  name: github
  # Replace this with the Github repo you want to make change
  repo: github-username/target-github-repo
  # This site_domain and base_url are sudo domains

>   # Replace site_domain with targetDomain pulumi stack configuration inside cms/infrastructure folder

Reviewer 2: I don't understand this comment (and the one on the next line too). Could you please take a look at this and perhaps rephrase it?

  # Replace base_url with targetDomain pulumi stack configuration inside cms-oauth/infrastructure folder

>   site_domain: https://some-cms-domain.pulumi-demos.com

Reviewer 2: ```suggestion
  site_domain: https://some-cms-domain.com
```

>   base_url: https://some-oauth-domain.pulumi-demos.com

Reviewer 2: ```suggestion
  base_url: https://some-oauth-domain.com
```

```

The site_domain is the URL of the CMS and the base_url is the URL of the OAuth Server . They should be the same as the `targetDomain` variable we set in the stack configuration of both CMS and OAuth Server.

## In the End

> Congratulations on building both CMS web application, the OAuth client-server, and deploy the infrastructure! Now if everything works perfectly, you would see the "Login with Github" button shown and click it would redirect to the Github login pages.

Reviewer 2: ```suggestion
Congratulations on building both CMS web application, the OAuth client-server, and also deploying the infrastructure! Now if everything works perfectly, you should see the **Login with Github** button and clicking it would redirect to the GitHub login page.
```

![OAuth Client Server](./oauth-client-server.jpg)

> People with correct access could now use and see the CMS.

Reviewer 2: ```suggestion
People with correct access can now use the CMS.
```

![CMS](./cms.jpg)

All codes of CMS and OAuth Client Server are available in [Pulumi's example repositories](https://github.com/pulumi/examples)'s [aws-ts-netlify-cms-and-oauth](https://github.com/pulumi/examples/tree/master/aws-ts-netlify-cms-and-oauth/cms-oauth).

---

Special Thanks to:

- Tony Alves @talves for providing [template](https://github.com/ADARTA/netlify-cms-react-example) to separate CMS as a stand-alone React App
- Igor Kuznetsov @igk1972 for providing [go code example](https://github.com/igk1972/netlify-cms-oauth-provider-go) for providing the OAuth Client-Server source code
- Paul Stack @stack72 for developing [Pulumi example of deploying Dockerized App using ECS Fargate](https://github.com/pulumi/examples/tree/master/aws-ts-hello-fargate) and [Pulumi example of deploying the static website on AWS](https://github.com/pulumi/examples/tree/master/aws-ts-static-website)

> - Every Pulumi folks have ever helped me out about this~

Reviewer 2: ```suggestion
- Everyone at Pulumi who helped me out with this
```
